### DISCUSSION TITLE
Pinterest: unable to search pins.

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#179] [Help Me!] [answered]
Gabby Tobermann 

Other help tickets: Yes.  https://github.com/nss-evening-cohort-14/evening-client-side/discussions/170 This one also deals with the search bar but it is not quite the same issue. 

Issue: I am able to search the boards in the search bar. The matching board is printing to the DOM. I applied the same logic to the pins, but I am getting an error an Uncaught TypeError:  Cannot read property 'toLowerCase' of undefined in my searchPins promise. It seems to not recognize the pin.title in my if statement (searchPins in pinData). I thought there might be something wrong with how I wrote my if statement in my domEvents, because I am calling searchBoards() and then searchPins(), but when I comment out searchBoards, the error is still there. 

errors: 
- pinData:
pinData.js?9601:61 Uncaught (in promise) TypeError: Cannot read property 'toLowerCase' of undefined
    at eval (pinData.js?9601:61)
    at Array.filter (<anonymous>)
    at eval (pinData.js?9601:58)
eval @ pinData.js?9601:61
eval @ pinData.js?9601:58
Promise.then (async)
eval @ navEvents.js?d0f4:21

-domEvents:
TypeError: Cannot read property 'requestContent' of undefined

Google links: 
I am not sure what to google, so I google the errors. 
https://stackoverflow.com/questions/62385133/typeerror-cannot-read-property-requestcontent-of-undefined-hmr-angular
https://community.hubspot.com/t5/APIs-Integrations/Getting-TypeError-Cannot-read-property-requestContent-of/td-p/315821
I also took another look at the docs on promises. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

What I have tried: 
- commented out searchBoards, so that I would only be searching pins, but the error is still there. 
- placed a debugger in my if statement in searchPins. When I hover over pin.title it's giving me the correct value, even though this is where the error tells me it is undefined. When I hover searchInput, it also gives me the correct value (they match).
- compared searchBoards() and searchPins(), they are identical. 
- checked my sample data to make sure it has a key of title, it does. 
- I console.warn(pin) in the filter method, all the pins are logged, including the one I searched for, with the correct title.
- I console.warn(filterValue) which is my filtered array, and it is returning the correct value. 

I'm probably overlooking something, but if anyone can take a look, I'd appreciate it! 

Links: 
pinData: https://github.com/Gabrielle-Tobermann/ASSIGNMENT-Firebase-Pinterest/blob/search-bar/src/javascripts/helpers/data/pinData.js
navEvents: https://github.com/Gabrielle-Tobermann/ASSIGNMENT-Firebase-Pinterest/blob/search-bar/src/javascripts/helpers/navEvents.js

**Comment by drteresavasquez** [comment]
You should only be lower casing the input and the values so if you're getting undefined, it may be in your input.

**Reply by drteresavasquez** [reply]
That's me guessing since you didn't provide a link to the file in your code.

**Reply by Gabrielle-Tobermann** [reply]
I'm sorry I always forget. I've added links.

**Comment by drteresavasquez** [comment]
In search pins, just use the filter to compare if the title includes the search value. @chiestroud got this working. See if she has some bandwidth to help or check out her code.

**Reply by Gabrielle-Tobermann** [reply]
Okay, I'll take a look at it. Thanks!

**Reply by drteresavasquez** [reply]
Also, @hmparsons-sde and @jrobinson0529 were able to get theirs working as well in a similar approach.

**Reply by Gabrielle-Tobermann** [reply]
So I rewrote my searchPins function, but I am still getting the same error. But with a debugger and console.warn(), I can see that the filterValue and the searchInput match though.

**Reply by chiestroud** [reply]
Gabby, I used two different functions like you, which are searchPin and searchBoard. In my case, I used Promise.all that have these two functions inside. And after that, in the navEvents, I used showPin & showBoard functions to run whenever enter key is pressed. As long as searchPin and searchBoard are working separately, you should be good. In my case, I was printing showPin & showBoard in the same div so didn't work, but after giving them separate div to print, it worked. I hope it helps. I will take more look at your code if this doesn't help :) Good luck! I was hitting my head to the wall yesterday with this!

**Reply by Gabrielle-Tobermann** [reply]
@chiestroud, thank you so much! Now I am able to search the pins also. It turns out one of my pins was "undefined" which is why none of my pins would work. I created two seperate containers, but for some reason my functions are still overriding each other... When my searchPins is before searchBoards, I can only search pins, and vice versa... 
NavEvents: https://github.com/Gabrielle-Tobermann/ASSIGNMENT-Firebase-Pinterest/blob/search-bar/src/javascripts/helpers/navEvents.js
BoardPinData: https://github.com/Gabrielle-Tobermann/ASSIGNMENT-Firebase-Pinterest/blob/search-bar/src/javascripts/helpers/data/boardPinData.js
domBuilder: https://github.com/Gabrielle-Tobermann/ASSIGNMENT-Firebase-Pinterest/blob/search-bar/src/javascripts/components/domBuilder.js

I am going through all of my code to make sure I am targetting the correct containers.

**Reply by hmparsons-sde** [reply]
Sorry! I didn't see this last night! glad you got it working :)

**Reply by Gabrielle-Tobermann** [reply]
No problem! It's not quite working but I'm getting there lol

**Comment by Gabrielle-Tobermann** [comment, marked as answer]
Update: It's kind of working, but I am reverting back to just searching the boards. I spent hours trying to make this work with the promise.all and extra containers, and I was getting very inconsistent results. I decided it would be safer to revert back to just searching the boards, without the promise.all(), extra components or containers. So I got back to that point, my boards were printing. I decided to try one more time to search the pins (I just can't let it go...) So I added my searchPins functions in my navEvents, and it works sometimes, but not always.  My original problem was that I had one pin that was undefined, so it was unable to search any pins. Now that I got rid of that pin, it works-ish. Sometimes it finds the correct pin or board, but other times it won't return anything. For the time being, since we have to present on Saturday, I will keep it to just searching the boards and continue to investigate for searching both. Not sure why the promise.all() with several containers wasn't working, but I think I had just made it too complicated. 

Lessons learned: 
- Stop obsessing over a stupid search bar! 
- Sometimes it's better to revert to where you know your code is working. 
- Make sure your data isn't wack... without that undefined pin, I would've saved hours of work. 

Thanks @chiestroud and @drteresavasquez for your help :)
